operator/watchers/cilium_podippool.go

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"context"
+
+	cilium_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+)
+
+// PooledAllocatorProvider defines the functions of IPAM provider front-end which additionally allow
+// definition of IP pools at runtime.
+// This is implemented by e.g. pkg/ipam/allocator/multipool
+type PooledAllocatorProvider interface {
+	UpsertPool(ctx context.Context, pool *cilium_v2alpha1.CiliumPodIPPool) error
+	DeletePool(ctx context.Context, pool *cilium_v2alpha1.CiliumPodIPPool) error
+}
+
+func StartIPPoolAllocator(
```

Comment:
How hard would it be to implement this as a hive module instead? I suppose we'd need to pull out `PooledAllocatorProvider` first from the legacy code plus make `alloc.Start` depend on `IPPoolAllocator` to make sure it's started first?

Reply:
I initially thought about implementing this as a hive cell. Unfortunately, it would require lifting quite a bit of IPAM code around. For example, I think we cannot make `alloc.Start` depend on `IPPoolAllocator` unconditionally, since only once specific implementation (namely `ipam/allocator/multipool.Allocator`) requires the pooled allocator, while all the others don't. Initially, I wanted to include `IPPoolAllocator` into `ipam/allocator.AllocatorProvider`, but IIRC that lead to nasty cyclic dependencies. Also, given this PR is already quite large and should make it into the tree for 1.14, I didn't want to include additional refactoring.

Admittedly it's not ideal to introduce new code in a non-modular fashion, but based on the above I'd defer addressing modularization of this code together with the other IPAM bits to a follow-up PR. I'll discuss the specifics with @gandro and @cilium/sig-foundations in a Slack thread or GH issue. Hope that's OK?

Reply:
Yeah, given the feature freeze coming up in a week, I don't think it's realistic to tackle a larger refactor at the moment unfortunately :crying_cat_face: 

The upside is that this is very little and simple code, so it should not make any future refactor more complicated